### PR TITLE
Implemented Shuffle Function for Beacon State

### PR DESCRIPTION
**Pull request by terenc3t**
To kick off on the beacon state transition function, we first implement a pure function to shuffle validator indices. This function simply shuffles an array of indices based on a random seed. For example:
```
validatorCount := 5
preShuffle := [0,1,2,3,4]
postShuffle: = [1,0,4,2,3]
```
Later, `getAttesters` and `getProposer` functions will use this shuffle function to return the list of validators.

References:
https://notes.ethereum.org/SCIg8AH5SA-O4C1G1LYZHQ?view#Helper-functions
https://github.com/ethereum/beacon_chain/blob/master/beacon_chain/state/state_transition.py#L43

### CODE REVIEW

File: beacon-chain/blockchain/core.go [hunk 1]
```diff
@@ -85,3 +89,32 @@ func (b *BeaconChain) persist() error {
 	}
 	return b.db.Put([]byte(stateLookupKey), encodedState)
 }
+
+// shuffle returns a list of pseudorandomly sampled
+// indices to use to select attesters and proposers.
+func shuffle(seed common.Hash, validatorCount int) ([]int, error) {
+	if validatorCount >= params.MaxValidators {
```
Comment (rawfalafel) [inline beacon-chain/blockchain/core.go]: should this be >, not >=?

File: beacon-chain/blockchain/core.go [hunk 2]
```diff
@@ -85,3 +89,32 @@ func (b *BeaconChain) persist() error {
...
+		return nil, errors.New("Validator count has exceeded MaxValidator Count")
+	}
+
+	// construct a list of indices up to MaxValidators
+	validatorList := make([]int, validatorCount)
+	for i := range validatorList {
+		validatorList[i] = i
+	}
+
+	hashSeed, _ := blake2s.New256(seed[:])
```
Comment (rawfalafel) [inline beacon-chain/blockchain/core.go]: is there a reason why you're not processing the error?

File: beacon-chain/blockchain/core.go [hunk 3]
```diff
@@ -85,3 +89,32 @@ func (b *BeaconChain) persist() error {
...
+	hashSeedByte := hashSeed.Sum(nil)
+
+	// shuffle stops at the second to last index
+	for i := 0; i < validatorCount-1; i++ {
+		// convert every 3 bytes to random number, replace validator index with that number
+		for j := 0; j+3 < len(hashSeedByte); j = j + 3 {
```
Comment (rawfalafel) [inline beacon-chain/blockchain/core.go]: (nitpick) `j+=3`

File: beacon-chain/blockchain/core_test.go [hunk 4]
```diff
@@ -119,3 +120,27 @@ func TestMutateCrystallizedState(t *testing.T) {
...
 }
+
+func TestFaultyShuffle(t *testing.T) {
+	if _, err := shuffle(common.Hash{'a'}, params.MaxValidators); err == nil {
+		t.Error("Shuffle should have failed when validator count exceeds MaxValidators")
+	}
+}
+
+func TestShuffle(t *testing.T) {
+	hash1 := common.BytesToHash([]byte{'a', 'b', 'c', 'd', 'e', 'f', 'g', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'a', 'b', 'c', 'd', 'e', 'f', 'g'})
```
Comment (rawfalafel) [inline beacon-chain/blockchain/core_test.go]: I would test `shuffle` for validator sizes n, where n%3=0, n%3=1, n%3=2

Comment (rawfalafel) [reply beacon-chain/blockchain/core_test.go]: i take this back. you're iterating every three bytes of the hash :sweat_smile:

**Comment by terenc3t** [comment]
@rawfalafel thanks for the feedback, everything's fixed. Ready again!

**Pull request merged**